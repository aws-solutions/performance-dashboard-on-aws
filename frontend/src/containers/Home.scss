.usa-search__submit-text {
    vertical-align: text-top !important;
}
