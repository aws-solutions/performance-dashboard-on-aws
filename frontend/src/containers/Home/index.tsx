import React, { useState, useEffect } from 'react';
import BadgerService from '../../services/badger-service';
import HomepageLayout from '../../layouts/Homepage';
import DashboardList from '../../components/DashboardList';

function Home() {
  const [dashboards, setDashboards] = useState([]);
  useEffect(() => {
    fetchDashboards();
  }, []);

  async function fetchDashboards() {
    const data = await BadgerService.fetchDashboards();
    setDashboards(data);
  }

  return (
    <HomepageLayout
      title="Performance Dashboard"
      subtitle="Find out how government services are performing and how we define our priorities to serve our citizens."
    >
      <DashboardList dashboards={dashboards} />
    </HomepageLayout>
  );
}

export default Home;
