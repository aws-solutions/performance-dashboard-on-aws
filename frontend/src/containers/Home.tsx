/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { useHistory } from "react-router-dom";
import { usePublicHomepage } from "../hooks";
import { useTranslation } from "react-i18next";
import UtilsService from "../services/UtilsService";
import Accordion from "../components/Accordion";
import Search from "../components/Search";
import { PublicDashboard, LocationState, PublicTopicArea, Dashboard } from "../models";
import Spinner from "../components/Spinner";
import MarkdownRender from "../components/MarkdownRender";
import CardGroup from "../components/CardGroup";
import dayjs from "dayjs";
import LocalizedFormat from "dayjs/plugin/localizedFormat";
import "./Home.scss";

dayjs.extend(LocalizedFormat);

function Home() {
    const { homepage, loading } = usePublicHomepage();
    const { t } = useTranslation();
    const history = useHistory<LocationState>();
    const { Card, CardBody } = CardGroup;

    const onSearch = (query: string) => {
        history.push("/public/search?q=" + query);
    };

    const onClear = () => {
        history.push("/");
    };

    const topicareas = UtilsService.groupByTopicArea(homepage.dashboards);

    // Sort topic areas and their dashboards in ascending alphabetical order.
    topicareas.sort((a: PublicTopicArea, b: PublicTopicArea) => a.name.localeCompare(b.name));
    for (let topicArea of topicareas) {
        topicArea.dashboards?.sort((a: PublicDashboard, b: PublicDashboard) =>
            a.name.localeCompare(b.name),
        );
    }

    const getDashboardLink = (dashboard: PublicDashboard | Dashboard): string => {
        const link = dashboard.friendlyURL ? dashboard.friendlyURL : dashboard.id;
        return link;
    };

    return loading ? (
        <Spinner
            style={{
                position: "fixed",
                top: "30%",
                left: "50%",
            }}
            label={t("LoadingSpinnerLabel")}
        />
    ) : (
        <div className="usa-prose">
            <div className="grid-row">
                <div className="grid-col-12 tablet:grid-col-8">
                    <h1 className="font-sans-3xl line-height-sans-2 margin-top-2">
                        {homepage.title}
                    </h1>
                    <MarkdownRender
                        className="font-sans-lg usa-prose"
                        source={homepage.description}
                    />
                </div>
            </div>
            <div className="grid-row">
                <div className="grid-col-12 padding-y-3 usa-prose">
                    <Search
                        id="search"
                        onSubmit={onSearch}
                        size="big"
                        onClear={onClear}
                        query=""
                        results={homepage.dashboards.length}
                        wide
                    />
                </div>
            </div>
            <div className="grid-row">
                <div className="grid-col-12 usa-prose">
                    <Accordion>
                        {topicareas.map((topicarea) => (
                            <Accordion.Item
                                id={topicarea.id}
                                key={topicarea.id}
                                title={`${topicarea.name} (${topicarea.dashboards?.length ?? 0})`}
                                hidden={true}
                            >
                                <ul className="usa-card-group">
                                    {topicarea?.dashboards?.map((dashboard) => (
                                        <Card
                                            id={dashboard.id}
                                            title={dashboard.name}
                                            col={4}
                                            link={getDashboardLink(dashboard)}
                                        >
                                            <CardBody>
                                                <p className="text-base text-italic margin-bottom-2">
                                                    {`${t("LastUpdatedLabel")} ${dayjs(
                                                        dashboard.updatedAt,
                                                    ).fromNow()}`}
                                                </p>
                                                {dashboard.description && (
                                                    <p>
                                                        <MarkdownRender
                                                            source={dashboard.description}
                                                            className="font-sans-md usa-prose margin-top-0"
                                                        />
                                                    </p>
                                                )}
                                            </CardBody>
                                        </Card>
                                    ))}
                                </ul>
                            </Accordion.Item>
                        ))}
                    </Accordion>
                </div>
            </div>
        </div>
    );
}

export default Home;
