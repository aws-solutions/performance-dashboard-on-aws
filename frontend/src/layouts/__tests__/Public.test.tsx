/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import { Route, Router, MemoryRouter } from "react-router-dom";
import { createMemoryHistory } from "history";
import withPublicLayout from "../Public";
import Home from "../../containers/Home";

const history = createMemoryHistory();

describe("PublicLayout", () => {
    test("renders the component", async () => {
        const wrapper = render(
            <Router history={history}>
                <Route component={withPublicLayout(Home)} />
            </Router>,
            { wrapper: MemoryRouter },
        );
        expect(wrapper.container).toMatchSnapshot();
    });
});
