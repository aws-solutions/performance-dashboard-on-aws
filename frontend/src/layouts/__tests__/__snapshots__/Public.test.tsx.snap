// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicLayout renders the component 1`] = `
<div>
  <div
    class="usa-overlay"
  />
  <header
    aria-label="Header"
    class="border-base-lighter border-width-1px shadow-2 z-top position-sticky top-0 bg-white usa-header usa-header--basic"
    role="contentinfo"
  >
    <div
      class="usa-nav-container"
    >
      <div
        class="usa-navbar navbar-long"
      >
        <div>
          <a
            class="usa-skipnav"
            href="#main"
          >
            Skip to main content
          </a>
        </div>
        <div
          class="usa-logo"
          id="basic-logo"
        >
          <em
            class="usa-logo__text display-flex flex-align-center"
          >
            <div
              class="logo"
            >
              <img
                alt="Organization logo"
                data-testid="logoImage"
                src="logo.svg"
              />
            </div>
            <a
              aria-label="Home"
              href="/"
              title="Home"
            />
          </em>
        </div>
        <button
          class="usa-menu-btn bg-primary-dark"
        >
          Menu
        </button>
      </div>
      <nav
        aria-label="Primary navigation"
        class="usa-nav"
      >
        <button
          aria-label="Close Menu"
          class="usa-nav__close"
        >
          <svg
            aria-hidden="true"
            aria-label="Close Menu"
            class="svg-inline--fa fa-window-close fa-w-16 fa-lg "
            data-icon="window-close"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M464 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-83.6 290.5c4.8 4.8 4.8 12.6 0 17.4l-40.5 40.5c-4.8 4.8-12.6 4.8-17.4 0L256 313.3l-66.5 67.1c-4.8 4.8-12.6 4.8-17.4 0l-40.5-40.5c-4.8-4.8-4.8-12.6 0-17.4l67.1-66.5-67.1-66.5c-4.8-4.8-4.8-12.6 0-17.4l40.5-40.5c4.8-4.8 12.6-4.8 17.4 0l66.5 67.1 66.5-67.1c4.8-4.8 12.6-4.8 17.4 0l40.5 40.5c4.8 4.8 4.8 12.6 0 17.4L313.3 256l67.1 66.5z"
              fill="currentColor"
            />
          </svg>
        </button>
        <ul
          class="usa-nav__primary usa-accordion"
        >
          <li
            class="usa-nav__primary-item"
          >
            <a
              aria-label="Contact"
              href="/public/contact"
              title="Contact"
            >
              Contact
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </header>
  <main
    class="padding-y-3"
  >
    <div
      id="main"
      tabindex="-1"
    />
    <div
      class="grid-container"
    >
      <div
        class="usa-prose"
      >
        <div
          class="grid-row"
        >
          <div
            class="grid-col-12 tablet:grid-col-8"
          >
            <h1
              class="font-sans-3xl line-height-sans-2 margin-top-2"
            >
              Kingdom of Wakanda
            </h1>
            <div
              class="Markdown font-sans-lg usa-prose"
            >
              <p>
                Welcome to our dashboard
              </p>
            </div>
          </div>
        </div>
        <div
          class="grid-row"
        >
          <div
            class="grid-col-12 padding-y-3 usa-prose"
          >
            <form
              class="usa-search usa-search--big bigwide"
              role="search"
            >
              <label
                class="usa-sr-only"
                for="search"
              >
                Search
              </label>
              <input
                class="usa-input"
                id="search"
                name="query"
                type="search"
                value=""
              />
              <button
                class="usa-button usa-button--base"
                type="submit"
              >
                <span
                  class="usa-search__submit-text"
                >
                  Search
                </span>
              </button>
            </form>
            
          </div>
        </div>
        <div
          class="grid-row"
        >
          <div
            class="grid-col-12 usa-prose"
          >
            <div
              class="usa-accordion"
            >
              <h2
                class="usa-accordion__heading"
              >
                <button
                  aria-controls="123456789"
                  aria-expanded="false"
                  class="usa-accordion__button"
                >
                  Topic Area Bananas (1)
                </button>
              </h2>
              <div
                class="usa-accordion__content usa-prose"
                hidden=""
                id="123456789"
              >
                <ul>
                  <ul
                    class="usa-card-group"
                  >
                    <li
                      class="usa-card cursor-default tablet:grid-col-4"
                      id="card-abc"
                    >
                      <div
                        class="usa-card__container"
                      >
                        <header
                          class="usa-card__header"
                        >
                          <h2
                            class="usa-card__heading font-family-sans"
                          >
                            <a
                              class="usa-link action-link"
                              href="/abc"
                              target=""
                            >
                              <span>
                                Dashboard One
                              </span>
                            </a>
                          </h2>
                        </header>
                        <div
                          class="usa-card__body"
                        >
                          <p
                            class="text-base text-italic margin-bottom-2"
                          >
                            Last updated a few seconds ago
                          </p>
                        </div>
                      </div>
                    </li>
                  </ul>
                </ul>
              </div>
              <h2
                class="usa-accordion__heading"
              >
                <button
                  aria-controls="987654321"
                  aria-expanded="false"
                  class="usa-accordion__button"
                >
                  Topic Area Grapes (1)
                </button>
              </h2>
              <div
                class="usa-accordion__content usa-prose"
                hidden=""
                id="987654321"
              >
                <ul>
                  <ul
                    class="usa-card-group"
                  >
                    <li
                      class="usa-card cursor-default tablet:grid-col-4"
                      id="card-xyz"
                    >
                      <div
                        class="usa-card__container"
                      >
                        <header
                          class="usa-card__header"
                        >
                          <h2
                            class="usa-card__heading font-family-sans"
                          >
                            <a
                              class="usa-link action-link"
                              href="/xyz"
                              target=""
                            >
                              <span>
                                Dashboard Two
                              </span>
                            </a>
                          </h2>
                        </header>
                        <div
                          class="usa-card__body"
                        >
                          <p
                            class="text-base text-italic margin-bottom-2"
                          >
                            Last updated a few seconds ago
                          </p>
                        </div>
                      </div>
                    </li>
                  </ul>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;
